README.md
# Introduction

This plugin to help use bluetooth printer in Android/iOS. Support for text, image, add new line or 
line dashed and QR. 

## Usage
### Initialize
```dart
BluePrintPos bluePrintPos = BluePrintPos.instance; 
```

### Scan bluetooth printer
```dart
bluePrintPos.scan();
```

### Connect bluetooth printer
```dart
bluePrintPos.connect(device);
```

### Print Text
In method ```.addText(text, {size, style, alignment})``` you can modify size, style and alignment
```dart
ReceiptSectionText receiptText = ReceiptSectionText();
receiptText.addText('MY STORE', size: ReceiptTextSizeType.medium, style: ReceiptTextStyleType.bold);
```

### Print text left right
```dart
receiptText.addLeftRightText('Time', '04/06/21, 10:00');
```

### Print image
```dart
final ByteData logoBytes = await rootBundle.load('assets/logo.jpg');
receiptText.addImage(
  base64.encode(Uint8List.view(logoBytes.buffer)),
  width: 150,
);
```

### Add new line
```dart 
receiptText.addSpacer();
```

### Add new line with dash
```dart 
receiptText.addSpacer(useDashed: true);
```

### Batch Printing
```dart
await bluePrintPos.printReceiptText(
  receipt,
  batchPrintOptions: const BatchPrintOptions.perNContent(20),
);
```

## Getting Started
### Android

Change the ```minSdkVersion``` in **android/app/build.gradle** in 19
```groovy
android {
  defaultConfig {
     minSdkVersion 19
  }
}
```

Add permission for Bluetooth and access location in **android/app/src/main/AndroidManifest.xml**
```xml 
<uses-permission android:name="android.permission.BLUETOOTH" />
<uses-permission android:name="android.permission.BLUETOOTH_ADMIN" />
<uses-permission android:name="android.permission.ACCESS_COARSE_LOCATION"/>
``` 

### iOS
Add info key in **ios/Runner/Info.plist**
```dart
<key>NSBluetoothAlwaysUsageDescription</key>  
<string>Need BLE permission</string>  
<key>NSBluetoothPeripheralUsageDescription</key>  
<string>Need BLE permission</string>  
<key>NSLocationAlwaysAndWhenInUseUsageDescription</key>  
<string>Need Location permission</string>  
<key>NSLocationAlwaysUsageDescription</key>  
<string>Need Location permission</string>  
<key>NSLocationWhenInUseUsageDescription</key>  
<string>Need Location permission</string>
``` 


## Thanks to
Reference and dependencies create this plugin
- [blue_thermal_printer](https://pub.dev/packages/blue_thermal_printer)
- [esc_pos_utils](https://pub.dev/packages/esc_pos_utils)
- [esc_pos_utils_plus](https://pub.dev/packages/esc_pos_utils_plus)
- [flutter_blue](https://pub.dev/packages/flutter_blue)
- [image](https://pub.dev/packages/image)
- [qr_flutter](https://pub.dev/packages/qr_flutter)
- [webcontent_converter](https://pub.dev/packages/webcontent_converter/score)




CHANGELOG.md
## 0.1.3
- chore: bump fluetooth to 0.0.5 ([#25](https://github.com/iandis/blue_print_pos/pull/23)) 

## 0.1.2
- feat: add printer feature ([#23](https://github.com/iandis/blue_print_pos/pull/23))

## 0.1.1
- fix: fix bluetooth permission on Android 12 ([#21](https://github.com/iandis/blue_print_pos/pull/21))

## 0.1.0
- **BREAKING CHANGE** refactor: change bluetooth package to use Fluetooth ([#19](https://github.com/iandis/blue_print_pos/pull/19))

## 0.0.23
- fix: add `await` to print processes 

## 0.0.22
- fix: empty section sublist and calculation ([#15](https://github.com/iandis/blue_print_pos/pull/13))

## 0.0.21
- fix: remove receipt vertical padding ([#13](https://github.com/iandis/blue_print_pos/pull/13))

## 0.0.20
- fix: batch print calculations ([#11](https://github.com/iandis/blue_print_pos/pull/11))

## 0.0.19
- feat: batch print ([#9](https://github.com/iandis/blue_print_pos/pull/9))

## 0.0.18
- chore: change `qr_flutter` from `pub`-hosted to `github`-hosted ([#7]((https://github.com/iandis/blue_print_pos/pull/7)))

## 0.0.17
- feat: adjustable font size on Android ([#5](https://github.com/iandis/blue_print_pos/pull/5))

## 0.0.16
- feat(receipt): add spacing using pixels ([#3](https://github.com/iandis/blue_print_pos/pull/3))

## 0.0.15

- fix(android): fix bitmap size on high density device ([#1](https://github.com/iandis/blue_print_pos/pull/1))

## 0.0.14

- feat: change [flutter_blue](https://pub.dev/packages/flutter_blue) with [flutter_blue_plus](https://pub.dev/packages/flutter_blue_plus) ([!53](https://github.com/andriyoganp/blue_print_pos/pull/53))
- fix: paper size print receipt text not pass bug ([#52](https://github.com/andriyoganp/blue_print_pos/pull/52))

## 0.0.13

- Update to flutter SDK 2.10.3, dart 2.16
- Change targetSDK to 32
- Fix NumberFormatException because null or empty
- Update log in debug only
- Fix CSS text-align in right

## 0.0.12

- Fix fail build on Android

## 0.0.11

- Provide default value on duration, only Android

## 0.0.10

- Set default duration delay in iOS
- Remove max width in method **.addImage**

## 0.0.9

- Add parameter **duration** for delay
- Add parameter **paperSize** to change paper width

## 0.0.8

- Add native code to convert content image

## 0.0.7

- Add parameter option to generate image as raster

## 0.0.6

- Add [webcontent_converter](https://pub.dev/packages/webcontent_converter) in library to make
  compatible in nullsafety

## 0.0.5

- Generate bytes with image raster enhancement

## 0.0.4

- Remove restriction service scanning iOS

## 0.0.3

- Decrease text small size into 0.8em

## 0.0.2

- Add documentation of code
- Add customization size for left-right text
- Change initialization with singleton

## 0.0.1

- Available print text, image, add new line, line dashed
- Support in Android and iOS


lib/blue_print_pos.dart
import 'dart:developer';
import 'dart:io';
import 'dart:typed_data';
import 'dart:ui';

import 'package:blue_print_pos/models/models.dart';
import 'package:blue_print_pos/receipt/receipt_section_text.dart';
import 'package:esc_pos_utils_plus/esc_pos_utils.dart';
import 'package:fluetooth/fluetooth.dart';
import 'package:flutter/services.dart';
import 'package:image/image.dart' as img;
import 'package:qr_flutter/qr_flutter.dart';

export 'package:esc_pos_utils_plus/esc_pos_utils.dart' show PaperSize;
export 'package:fluetooth/fluetooth.dart' show FluetoothDevice;
export 'models/models.dart';
export 'receipt/receipt.dart';

class BluePrintPos {
  static final BluePrintPos _instance = BluePrintPos();

  static BluePrintPos get instance => _instance;

  static const MethodChannel _channel = MethodChannel('blue_print_pos');

  /// State to get bluetooth is connected
  bool _isConnected = false;

  /// Getter value [_isConnected]
  bool get isConnected => _isConnected;

  FluetoothDevice? _selectedDevice;

  /// Selected device after connecting
  FluetoothDevice? get selectedDevice => _selectedDevice;

  /// return bluetooth device list, handler Android and iOS in [BlueScanner]
  Future<List<FluetoothDevice>> scan() {
    return Fluetooth().getAvailableDevices();
  }

  /// When connecting, reassign value [selectedDevice] from parameter [device]
  /// and if connection time more than [timeout]
  /// will return [ConnectionStatus.timeout]
  /// When connection success, will return [ConnectionStatus.connected]
  Future<ConnectionStatus> connect(
    FluetoothDevice device, {
    Duration timeout = const Duration(seconds: 5),
  }) async {
    try {
      final FluetoothDevice fDevice =
          await Fluetooth().connect(device.id).timeout(timeout);
      _selectedDevice = fDevice;
      _isConnected = true;
      return Future<ConnectionStatus>.value(ConnectionStatus.connected);
    } on Exception catch (error) {
      log('$runtimeType - Error $error');
      _isConnected = false;
      _selectedDevice = null;
      return Future<ConnectionStatus>.value(ConnectionStatus.timeout);
    }
  }

  /// To stop communication between bluetooth device and application
  Future<ConnectionStatus> disconnect() async {
    await Fluetooth().disconnect();
    _isConnected = false;
    _selectedDevice = null;
    return ConnectionStatus.disconnect;
  }

  /// This method only for print text
  /// value and styling inside model [ReceiptSectionText].
  /// [feedCount] to create more space after printing process done
  ///
  /// [useCut] to cut printing process
  ///
  /// [duration] the delay duration before converting the html to bytes.
  /// defaults to 0.
  ///
  /// [textScaleFactor] the text scale factor (must be > 0 or null).
  /// note that this currently only works on Android.
  /// defaults to system's font settings.
  /// 
  /// [batchPrintOptions] to print each [ReceiptSectionText]'s content in batch.
  /// defaults to [BatchPrintOptions.full].
  Future<void> printReceiptText(
    ReceiptSectionText receiptSectionText, {
    int feedCount = 0,
    bool useCut = false,
    bool useRaster = false,
    double duration = 0,
    PaperSize paperSize = PaperSize.mm58,
    double? textScaleFactor,
    BatchPrintOptions? batchPrintOptions,
  }) async {
    final int contentLength = receiptSectionText.contentLength;

    final BatchPrintOptions batchOptions =
        batchPrintOptions ?? BatchPrintOptions.full;

    final Iterable<List<Object>> startEndIter =
        batchOptions.getStartEnd(contentLength);

    for (final List<Object> startEnd in startEndIter) {
      final ReceiptSectionText section = receiptSectionText.getSection(
        startEnd[0] as int,
        startEnd[1] as int,
      );
      final bool isEndOfBatch = startEnd[2] as bool;
      final Uint8List bytes = await contentToImage(
        content: section.getContent(),
        duration: duration,
        textScaleFactor: textScaleFactor,
      );
      final List<int> byteBuffer = await _getBytes(
        bytes,
        paperSize: paperSize,
        feedCount: isEndOfBatch ? feedCount : batchOptions.feedCount,
        useCut: isEndOfBatch ? useCut : batchOptions.useCut,
        useRaster: useRaster,
      );
      await _printProcess(byteBuffer);
      log(
        'start: ${startEnd[0]} end: ${startEnd[1]}',
        name: 'BluePrintPos.printReceiptText',
      );

      if (batchOptions.delay != Duration.zero &&
          !batchOptions.delay.isNegative) {
        await Future<void>.delayed(batchOptions.delay);
      }
    }
  }

  /// This method only for print image with parameter [bytes] in List<int>
  /// define [width] to custom width of image, default value is 120
  /// [feedCount] to create more space after printing process done
  /// [useCut] to cut printing process
  Future<void> printReceiptImage(
    List<int> bytes, {
    int width = 120,
    int feedCount = 0,
    bool useCut = false,
    bool useRaster = false,
    PaperSize paperSize = PaperSize.mm58,
  }) async {
    final List<int> byteBuffer = await _getBytes(
      bytes,
      customWidth: width,
      feedCount: feedCount,
      useCut: useCut,
      useRaster: useRaster,
      paperSize: paperSize,
    );
    await _printProcess(byteBuffer);
  }

  /// This method only for print QR, only pass value on parameter [data]
  /// define [size] to size of QR, default value is 120
  /// [feedCount] to create more space after printing process done
  /// [useCut] to cut printing process
  Future<void> printQR(
    String data, {
    int size = 120,
    int feedCount = 0,
    bool useCut = false,
  }) async {
    final List<int> byteBuffer = await _getQRImage(data, size.toDouble());
    await printReceiptImage(
      byteBuffer,
      width: size,
      feedCount: feedCount,
      useCut: useCut,
    );
  }

  /// Reusable method for print text, image or QR based value [byteBuffer]
  /// Handler Android or iOS will use method writeBytes from ByteBuffer
  /// But in iOS more complex handler using service and characteristic
  Future<void> _printProcess(List<int> byteBuffer) async {
    try {
      if (!await Fluetooth().isConnected) {
        _isConnected = false;
        _selectedDevice = null;
        return;
      }
      await Fluetooth().sendBytes(byteBuffer);
    } on Exception catch (error) {
      log('$runtimeType - Error $error');
    }
  }

  /// This method to convert byte from [data] into as image canvas.
  /// It will automatically set width and height based [paperSize].
  /// [customWidth] to print image with specific width
  /// [feedCount] to generate byte buffer as feed in receipt.
  /// [useCut] to cut of receipt layout as byte buffer.
  Future<List<int>> _getBytes(
    List<int> data, {
    PaperSize paperSize = PaperSize.mm58,
    int customWidth = 0,
    int feedCount = 0,
    bool useCut = false,
    bool useRaster = false,
  }) async {
    List<int> bytes = <int>[];
    final CapabilityProfile profile = await CapabilityProfile.load();
    final Generator generator = Generator(paperSize, profile);
    final img.Image _resize = img.copyResize(
      img.decodeImage(data)!,
      width: customWidth > 0 ? customWidth : paperSize.width,
    );
    if (useRaster) {
      bytes += generator.imageRaster(_resize);
    } else {
      bytes += generator.image(_resize);
    }
    if (feedCount > 0) {
      bytes += generator.feed(feedCount);
    }
    if (useCut) {
      bytes += generator.cut();
    }
    return bytes;
  }

  /// Handler to generate QR image from [text] and set the [size].
  /// Using painter and convert to [Image] object and return as [Uint8List]
  Future<Uint8List> _getQRImage(String text, double size) async {
    try {
      final Image image = await QrPainter(
        data: text,
        version: QrVersions.auto,
        gapless: false,
        color: const Color(0xFF000000),
        emptyColor: const Color(0xFFFFFFFF),
      ).toImage(size);
      final ByteData? byteData =
          await image.toByteData(format: ImageByteFormat.png);
      assert(byteData != null);
      return byteData!.buffer.asUint8List();
    } on Exception catch (exception) {
      print('$runtimeType - $exception');
      rethrow;
    }
  }

  /// Converts HTML content to bytes
  ///
  /// [content] the html text
  ///
  /// [duration] the delay duration before converting the html to bytes.
  /// defaults to 0.
  ///
  /// [textScaleFactor] the text scale factor (must be > 0 or null).
  /// note that this currently only works on Android.
  /// defaults to system's font settings.
  static Future<Uint8List> contentToImage({
    required String content,
    double duration = 0,
    double? textScaleFactor,
  }) async {
    assert(
      textScaleFactor == null || textScaleFactor > 0,
      '`textScaleFactor` must be either null or more than zero.',
    );
    final Map<String, dynamic> arguments = <String, dynamic>{
      'content': content,
      'duration': Platform.isIOS ? 2000 : duration,
      'textScaleFactor': textScaleFactor,
    };
    Uint8List results = Uint8List.fromList(<int>[]);
    try {
      results = await _channel.invokeMethod('contentToImage', arguments) ??
          Uint8List.fromList(<int>[]);
    } on Exception catch (e) {
      log('[method:contentToImage]: $e');
      throw Exception('Error: $e');
    }
    return results;
  }
}


example/lib/main.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:blue_print_pos/blue_print_pos.dart';
import 'package:blue_print_pos/models/models.dart';
import 'package:blue_print_pos/receipt/receipt.dart';
import 'package:flutter/material.dart';
import 'package:flutter/services.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatefulWidget {
  @override
  _MyAppState createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  final BluePrintPos _bluePrintPos = BluePrintPos.instance;
  List<FluetoothDevice> _blueDevices = <FluetoothDevice>[];
  FluetoothDevice? _selectedDevice;
  bool _isLoading = false;
  int _loadingAtIndex = -1;

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      home: Scaffold(
        appBar: AppBar(
          title: const Text('Blue Print Pos'),
        ),
        body: SafeArea(
          child: _isLoading && _blueDevices.isEmpty
              ? const Center(
                  child: CircularProgressIndicator(
                    valueColor: AlwaysStoppedAnimation<Color>(Colors.blue),
                  ),
                )
              : _blueDevices.isNotEmpty
                  ? SingleChildScrollView(
                      child: Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: <Widget>[
                          Column(
                            children: List<Widget>.generate(_blueDevices.length,
                                (int index) {
                              return Row(
                                children: <Widget>[
                                  Expanded(
                                    child: GestureDetector(
                                      onTap: _blueDevices[index].id ==
                                              (_selectedDevice?.id ?? '')
                                          ? _onDisconnectDevice
                                          : () => _onSelectDevice(index),
                                      child: Padding(
                                        padding: const EdgeInsets.all(8.0),
                                        child: Column(
                                          crossAxisAlignment:
                                              CrossAxisAlignment.start,
                                          children: <Widget>[
                                            Text(
                                              _blueDevices[index].name,
                                              style: TextStyle(
                                                color: _selectedDevice?.id ==
                                                        _blueDevices[index].id
                                                    ? Colors.blue
                                                    : Colors.black,
                                                fontSize: 20,
                                                fontWeight: FontWeight.w500,
                                              ),
                                            ),
                                            Text(
                                              _blueDevices[index].id,
                                              style: TextStyle(
                                                color: _selectedDevice?.id ==
                                                        _blueDevices[index].id
                                                    ? Colors.blueGrey
                                                    : Colors.grey,
                                                fontSize: 14,
                                                fontWeight: FontWeight.w500,
                                              ),
                                            ),
                                          ],
                                        ),
                                      ),
                                    ),
                                  ),
                                  if (_loadingAtIndex == index && _isLoading)
                                    Container(
                                      height: 24.0,
                                      width: 24.0,
                                      margin: const EdgeInsets.only(right: 8.0),
                                      child: const CircularProgressIndicator(
                                        valueColor:
                                            AlwaysStoppedAnimation<Color>(
                                          Colors.blue,
                                        ),
                                      ),
                                    ),
                                  if (!_isLoading &&
                                      _blueDevices[index].id ==
                                          (_selectedDevice?.id ?? ''))
                                    TextButton(
                                      onPressed: _onPrintReceipt,
                                      child: Container(
                                        color: _selectedDevice == null
                                            ? Colors.grey
                                            : Colors.blue,
                                        padding: const EdgeInsets.all(8.0),
                                        child: const Text(
                                          'Test Print',
                                          style: TextStyle(color: Colors.white),
                                        ),
                                      ),
                                      style: ButtonStyle(
                                        backgroundColor: MaterialStateProperty
                                            .resolveWith<Color>(
                                          (Set<MaterialState> states) {
                                            if (states.contains(
                                                MaterialState.pressed)) {
                                              return Theme.of(context)
                                                  .colorScheme
                                                  .primary
                                                  .withOpacity(0.5);
                                            }
                                            return Theme.of(context)
                                                .primaryColor;
                                          },
                                        ),
                                      ),
                                    ),
                                ],
                              );
                            }),
                          ),
                        ],
                      ),
                    )
                  : Center(
                      child: Column(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: const <Widget>[
                          Text(
                            'Scan bluetooth device',
                            style: TextStyle(fontSize: 24, color: Colors.blue),
                          ),
                          Text(
                            'Press button scan',
                            style: TextStyle(fontSize: 14, color: Colors.grey),
                          ),
                        ],
                      ),
                    ),
        ),
        floatingActionButton: FloatingActionButton(
          onPressed: _isLoading ? null : _onScanPressed,
          child: const Icon(Icons.search),
          backgroundColor: _isLoading ? Colors.grey : Colors.blue,
        ),
      ),
    );
  }

  Future<void> _onScanPressed() async {
    setState(() => _isLoading = true);
    _bluePrintPos.scan().then((List<FluetoothDevice> devices) {
      if (devices.isNotEmpty) {
        setState(() {
          _blueDevices = devices;
          _isLoading = false;
        });
      } else {
        setState(() => _isLoading = false);
      }
    });
  }

  void _onDisconnectDevice() {
    _bluePrintPos.disconnect().then((ConnectionStatus status) {
      if (status == ConnectionStatus.disconnect) {
        setState(() {
          _selectedDevice = null;
        });
      }
    });
  }

  void _onSelectDevice(int index) {
    setState(() {
      _isLoading = true;
      _loadingAtIndex = index;
    });
    final FluetoothDevice blueDevice = _blueDevices[index];
    _bluePrintPos.connect(blueDevice).then((ConnectionStatus status) {
      if (status == ConnectionStatus.connected) {
        setState(() => _selectedDevice = blueDevice);
      } else if (status == ConnectionStatus.timeout) {
        _onDisconnectDevice();
      } else {
        print('$runtimeType - something wrong');
      }
      setState(() => _isLoading = false);
    });
  }

  Future<void> _onPrintReceipt() async {
    /// Example for Print Image
    final ByteData logoBytes = await rootBundle.load(
      'assets/logo.jpg',
    );

    /// Example for Print Text
    final ReceiptSectionText receiptText = ReceiptSectionText();
    receiptText.addImage(
      base64.encode(Uint8List.view(logoBytes.buffer)),
      width: 150,
    );
    receiptText.addSpacer();
    receiptText.addText(
      'MY STORE',
      size: ReceiptTextSizeType.medium,
      style: ReceiptTextStyleType.bold,
    );
    receiptText.addText(
      'Black White Street, Jakarta, Indonesia',
      size: ReceiptTextSizeType.small,
    );
    receiptText.addSpacer(useDashed: true);
    receiptText.addLeftRightText('Time', '04/06/21, 10:00');
    receiptText.addSpacer(useDashed: true);
    receiptText.addLeftRightText(
      'Apple 1kg',
      'Rp30.000',
      leftStyle: ReceiptTextStyleType.normal,
      rightStyle: ReceiptTextStyleType.bold,
    );
    receiptText.addSpacer(useDashed: true);
    receiptText.addLeftRightText(
      'TOTAL',
      'Rp30.000',
      leftStyle: ReceiptTextStyleType.normal,
      rightStyle: ReceiptTextStyleType.bold,
    );
    receiptText.addSpacer(useDashed: true);
    receiptText.addLeftRightText(
      'Payment',
      'Cash',
      leftStyle: ReceiptTextStyleType.normal,
      rightStyle: ReceiptTextStyleType.normal,
    );
    receiptText.addSpacer(count: 2);

    await _bluePrintPos.printReceiptText(receiptText);

    /// Example for print QR
    await _bluePrintPos.printQR('www.google.com', size: 250);

    /// Text after QR
    final ReceiptSectionText receiptSecondText = ReceiptSectionText();
    receiptSecondText.addText('Powered by ayeee',
        size: ReceiptTextSizeType.small);
    receiptSecondText.addSpacer();
    await _bluePrintPos.printReceiptText(receiptSecondText, feedCount: 1);
  }
}
